types/wlr_cursor.c

```diff
@@ -330,11 +330,83 @@ static void handle_pointer_motion(struct wl_listener *listener, void *data) {
 	wlr_signal_emit_safe(&device->cursor->events.motion, event);
 }
 
+static void apply_output_transform(double *x, double *y,
+		enum wl_output_transform transform) {
+	double dx, dy;
+	double width = 1.0, height = 1.0;
+
+	switch (transform) {
+	case WL_OUTPUT_TRANSFORM_NORMAL:
+		dx = *x;
+		dy = *y;
+		break;
+	case WL_OUTPUT_TRANSFORM_90:
+		dx = *y;
+		dy = width - *x;
+		break;
+	case WL_OUTPUT_TRANSFORM_180:
+		dx = width - *x;
+		dy = height - *y;
+		break;
+	case WL_OUTPUT_TRANSFORM_270:
+		dx = height - *y;
+		dy = *x;
+		break;
+	case WL_OUTPUT_TRANSFORM_FLIPPED:
+		dx = width - *x;
+		dy = *y;
+		break;
+	case WL_OUTPUT_TRANSFORM_FLIPPED_90:
+		dx = height - *y;
+		dy = width - *x;
+		break;
+	case WL_OUTPUT_TRANSFORM_FLIPPED_180:
+		dx = *x;
+		dy = height - *y;
+		break;
+	case WL_OUTPUT_TRANSFORM_FLIPPED_270:
+		dx = *y;
+		dy = *x;
+		break;
+	}
+	*x = dx;
+	*y = dy;
+}
+
+
+/* get output based on mapped_output or input device coordinates */
+static struct wlr_output *get_output(struct wlr_cursor_device *cursor_device,
+									 struct wlr_input_device *input_device,
+									 double x, double y)
+{
+	if (cursor_device->mapped_output) {
+		return cursor_device->mapped_output;
+	}
+
+	struct wlr_cursor *cursor = cursor_device->cursor;
+	assert(cursor);
+	if (cursor->state->mapped_output) {
+		return cursor->state->mapped_output;
+	}
+
+	double lx, ly;
+	wlr_cursor_absolute_to_layout_coords(cursor_device->cursor, input_device,
+		x, y, &lx, &ly);
+	return wlr_output_layout_output_at(cursor->state->layout, lx, ly);
```

Comment:
It shouldn't try to be smart and guess the output based on coords IMHO.

Reply:
If I don't do that  I don't end up with an output at all and so no transformation can be done although it is rotated.

Reply:
Hmm, yeah, after thinking about it it's probably fine.

Reply:
>If I don't do that I don't end up with an output at all and so no transformation can be done although it is rotated.

Hmm, nah, in the end I really think we shouldn't do that. You should map the touch device to the output, otherwise it'll break support for multiple outputs.

Reply:
@emersion you mean map configuration wise or code wise?

Reply:
That's up to the compositor, so code wise. See https://github.com/swaywm/wlroots/issues/925#issuecomment-385175579 for a list of possible solutions. In rootston that would mean add some configuration.

The behaviour you implemented works with 1 output because the function always return the output. But it really doesn't make sense with two outputs. I just tested by removing the backend-wise mapping in the X11 backend and configuring one screen to be rotated, the cursor jumps when you move it. (If you try with the Wayland backend you need to disable hardware cursors)

Reply:
Please ignore my ignorance but which code did you exactly disable? If I run this PR (which includes the removal of X11/Wayland specific coordinates) I get the same behaviour as on master.

I know about #925, that's why I suggested to map via wlr_input_device.output_name. But that should probably be done in addition to the current code. IMHO special configuration should not be required for the simple case of a monitor with touch capabilities.